Change kubectl diff to exclude managedFields by default

brianpursley opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Changes kubectl diff to exclude managedFields by default.
Adds a new --show-managed-fields flag that allows you to include managed fields in the diff.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubectl/issues/1242

#### Special notes for your reviewer:
New Output:
```
diff -u -N /tmp/LIVE-4192522669/apps.v1.Deployment.nginx-test.foo /tmp/MERGED-3896013886/apps.v1.Deployment.nginx-test.foo
--- /tmp/LIVE-4192522669/apps.v1.Deployment.nginx-test.foo	2022-07-21 11:11:39.107811757 -0400
+++ /tmp/MERGED-3896013886/apps.v1.Deployment.nginx-test.foo	2022-07-21 11:11:39.111811814 -0400
@@ -6,7 +6,7 @@
     kubectl.kubernetes.io/last-applied-configuration: |
       {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"name":"foo","namespace":"nginx-test"},"spec":{"replicas":1,"selector":{"matchLabels":{"app":"foo"}},"template":{"metadata":{"labels":{"app":"foo"}},"spec":{"containers":[{"command":["sh","-c","echo Container bar is running! \u0026\u0026 sleep 9999999"],"image":"busybox:1.30","name":"bar"}]}}}}
   creationTimestamp: "2022-07-21T13:31:29Z"
-  generation: 1
+  generation: 2
   name: foo
   namespace: nginx-test
   resourceVersion: "38911353"
@@ -14,7 +14,7 @@
 spec:
   progressDeadlineSeconds: 600
   replicas: 1
-  revisionHistoryLimit: 10
+  revisionHistoryLimit: 11
   selector:
     matchLabels:
       app: foo
exit status 1
```

Old output:
```
diff -u -N /tmp/LIVE-2971090238/apps.v1.Deployment.nginx-test.foo /tmp/MERGED-4094261447/apps.v1.Deployment.nginx-test.foo
--- /tmp/LIVE-2971090238/apps.v1.Deployment.nginx-test.foo	2022-07-21 11:12:22.416433597 -0400
+++ /tmp/MERGED-4094261447/apps.v1.Deployment.nginx-test.foo	2022-07-21 11:12:22.420433655 -0400
@@ -6,13 +6,47 @@
     kubectl.kubernetes.io/last-applied-configuration: |
       {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"name":"foo","namespace":"nginx-test"},"spec":{"replicas":1,"selector":{"matchLabels":{"app":"foo"}},"template":{"metadata":{"labels":{"app":"foo"}},"spec":{"containers":[{"command":["sh","-c","echo Container bar is running! \u0026\u0026 sleep 9999999"],"image":"busybox:1.30","name":"bar"}]}}}}
   creationTimestamp: "2022-07-21T13:31:29Z"
-  generation: 1
+  generation: 2
   managedFields:
   - apiVersion: apps/v1
     fieldsType: FieldsV1
     fieldsV1:
       f:metadata:
         f:annotations:
+          f:deployment.kubernetes.io/revision: {}
+      f:status:
+        f:availableReplicas: {}
+        f:conditions:
+          .: {}
+          k:{"type":"Available"}:
+            .: {}
+            f:lastTransitionTime: {}
+            f:lastUpdateTime: {}
+            f:message: {}
+            f:reason: {}
+            f:status: {}
+            f:type: {}
+          k:{"type":"Progressing"}:
+            .: {}
+            f:lastTransitionTime: {}
+            f:lastUpdateTime: {}
+            f:message: {}
+            f:reason: {}
+            f:status: {}
+            f:type: {}
+        f:observedGeneration: {}
+        f:readyReplicas: {}
+        f:replicas: {}
+        f:updatedReplicas: {}
+    manager: kube-controller-manager
+    operation: Update
+    subresource: status
+    time: "2022-07-21T13:31:30Z"
+  - apiVersion: apps/v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:metadata:
+        f:annotations:
           .: {}
           f:kubectl.kubernetes.io/last-applied-configuration: {}
       f:spec:
@@ -49,41 +83,7 @@
             f:terminationGracePeriodSeconds: {}
     manager: kubectl-client-side-apply
     operation: Update
-    time: "2022-07-21T13:31:29Z"
-  - apiVersion: apps/v1
-    fieldsType: FieldsV1
-    fieldsV1:
-      f:metadata:
-        f:annotations:
-          f:deployment.kubernetes.io/revision: {}
-      f:status:
-        f:availableReplicas: {}
-        f:conditions:
-          .: {}
-          k:{"type":"Available"}:
-            .: {}
-            f:lastTransitionTime: {}
-            f:lastUpdateTime: {}
-            f:message: {}
-            f:reason: {}
-            f:status: {}
-            f:type: {}
-          k:{"type":"Progressing"}:
-            .: {}
-            f:lastTransitionTime: {}
-            f:lastUpdateTime: {}
-            f:message: {}
-            f:reason: {}
-            f:status: {}
-            f:type: {}
-        f:observedGeneration: {}
-        f:readyReplicas: {}
-        f:replicas: {}
-        f:updatedReplicas: {}
-    manager: kube-controller-manager
-    operation: Update
-    subresource: status
-    time: "2022-07-21T13:31:30Z"
+    time: "2022-07-21T15:12:22Z"
   name: foo
   namespace: nginx-test
   resourceVersion: "38911353"
@@ -91,7 +91,7 @@
 spec:
   progressDeadlineSeconds: 600
   replicas: 1
-  revisionHistoryLimit: 10
+  revisionHistoryLimit: 11
   selector:
     matchLabels:
       app: foo
```


#### Does this PR introduce a user-facing change?
```release-note
kubectl diff changed to ignore managed fields by default, and a new --show-managed-fields flag has been added to allow you to include managed fields in the diff
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

brianpursley commented:
/hold for review
/cc @ardaguclu what do you think of this change?

brianpursley commented:
I named the flag `--include-managed-fields` instead of `--show-managed-fields` because with diff, it is not guaranteed they will be shown, only that they would be included in the diff decision.  I wavered on whether to do this though because the flag is called `--show-managed-fields` in `kubectl get`

ardaguclu commented:
Thanks for the PR @brianpursley!

In my opinion, we may not need new flag to fix that problem;

- If this issue https://github.com/kubernetes/kubectl/issues/1242 is a regression caused by referenced PR(which I think it is), there should not be any changes needed because `diff` already shows expected results in older releases.
- If issue is not a regression and an inevitable change, there are different use cases we need to consider;

1. `kubectl apply -f foo.yaml` -> `kubectl diff -f foo.yaml`
Both of them are client side apply and there is no managedfields already. In some cases(like operators take ownership and apply SSA as in our case), there will be managedFields and I think it is a good thing to show these fields to users to warn that ownership changed, etc.
2. `kubectl apply -f --server-side foo.yaml` -> `kubectl diff -f foo.yaml`
3. `kubectl apply -f foo.yaml` -> `kubectl diff --server-side -f foo.yaml`
AFAIK, we don't support 2. and 3. or maybe I'm wrong(in any case, these are edge cases)
4. `kubectl apply --server-side -f foo.yaml` -> `kubectl diff --server-side -f foo.yaml`
I think, we should always show managed fields if there is any change in there. Because that will be a good indicator to the user about new fields, possible ownership changes, etc. and if we hide managed fields that will reduce the usability of diff command for SSA. 

In my opinion this new `include-managed-fields` flag should already be always to true in SSA and flag does not matter in client side apply(only usable for the case I mentioned: ownership changes). In addition to that, in the future, someday, SSA will be default and very likely this new flag will be deprecated?.

To sum up, if there is a regression that means `apply` also is affected and we need to fix that. If not, instead exposing a new flag, we can always show managed fields in SSA and always disable in client side apply maybe(or always show managed fields)?

brianpursley commented:
Thanks @ardaguclu.  I'll need to look at that other commit you referenced and understand it better.

Something I noticed when I looked at the live and merged files being compared is that the managed fields came back in a different order (reversed) in the live vs the merged, and that was the cause of the big diff in the example output I posted.

So it could also be an ordering problem and maybe it can be fixed if managed fields are always returned in a consistent order so that the diff doesn't pick up the reordering as a change.

ardaguclu commented:
@brianpursley you are right, always updating time fields ,s causing different ordering each time(and diff results). Because fields are sorted according to time field https://github.com/kubernetes/kubernetes/blob/862d256195adf3be5475b1a6935e5feb78f884a5/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go#L194

When I disable sorting manually, diff works as expected. I wonder can we handle that case _somehow_ in diff command.

ardaguclu commented:
Maybe the best option is like you said hiding altogether and only show them when user passes `include-managed-fields`. Because time field always will be changed and always will be in diff. WDYT?

brianpursley commented:
@ardaguclu thanks for looking into it further.  Your explanation makes sense to me.

I think it is preferable for diff to exclude managed fields by default.  Including them doesn't seem like helpful information to me.

I actually would be OK with always excluding the managed fields and not even having the `--include-managed-fields` flag, but was hesitant to go that far in case someone wanted to see them.  I could easily be convinced the flag isn't necessary though.

I guess we start without the flag and if there is demand for that capability, then add the flag to allow managed fields to be included.

ardaguclu commented:
> @ardaguclu thanks for looking into it further. Your explanation makes sense to me.
> 
> I think it is preferable for diff to exclude managed fields by default. Including them doesn't seem like helpful information to me.
> 
> I actually would be OK with always excluding the managed fields and not even having the `--include-managed-fields` flag, but was hesitant to go that far in case someone wanted to see them. I could easily be convinced the flag isn't necessary though.
> 
> I guess we start without the flag and if there is demand for that capability, then add the flag to allow managed fields to be included.

Thanks, maybe like you said we can directly disable managedfields and will expose this flag if there is any demand? but I'm fine both options.

Your PR also looks good to me but I wonder why none of integration tests did not fail and maybe we need one for managedField case?

KnVerey commented:
> I guess we start without the flag and if there is demand for that capability, then add the flag to allow managed fields to be included.

If the output as-is is always useless, then this makes sense to me. I don't see the sense in maintaining a flag for something that isn't helpful. If we do have a flag, I'd expect we'd need to find a way to fix the problem of the diffing essentially not working for managed fields before exposing it.

That said, since this is a behaviour change either way, I suggest we discuss this briefly at tomorrow's meeting.

apelisse commented:
I personally believe that diffing the managed fields is generally useful. The other mentioned error is definitely not a bug. On the other hand, I agree that the re-ordering of managers in this case is not super helpful. I'm curious if there's something we can do about that.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -583,6 +590,16 @@ func (d *Differ) Diff(obj Object, printer Printer) error {
 	return nil
 }
 
+func removeManagedFields(obj runtime.Object) {
```
Nit: can you do something similar to https://github.com/kubernetes/kubernetes/blob/610b7839a0bc2f9bafaca92930f8dd740f0ab46c/staging/src/k8s.io/cli-runtime/pkg/printers/managedfields.go#L34-L42

brianpursley replied:
Whoops, I just saw this comment, will make this change...

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -164,6 +165,7 @@ func NewCmdDiff(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 	usage := "contains the configuration to diff"
 	cmd.Flags().StringArray("prune-allowlist", []string{}, "Overwrite the default whitelist with <group/version/kind> for --prune")
 	cmd.Flags().Bool("prune", false, "Include resources that would be deleted by pruning. Can be used with -l and default shows all resources would be pruned")
+	cmd.Flags().BoolVar(&options.IncludeManagedFields, "include-managed-fields", options.IncludeManagedFields, "If true, include managed fields in the diff.")
```
As discussed during sig-cli call, change this to `--show-managed-fields` to be consistent with `kubectl get`.

brianpursley replied:
Flag name has been changed

brianpursley committed:
Change kubectl diff to exclude managedFields by default

Changes kubectl diff to exclude managedFields by default.
Adds a new --show-managed-fields flag that allows you to
include managed fields in the diff.

soltysh reviewed (approved):
/lgtm

soltysh commented:
/priority important-longterm
/triage accepted

soltysh commented:
/hold cancel
this was discussed during today's sig-cli and the comments above from me are part of that.
